### PR TITLE
Async batch predictions

### PR DESCRIPTION
**Pull request by jkrukowski**
Should partially resolve https://github.com/argmaxinc/WhisperKit/issues/97

- added audio file batch processing
- added `concurrentWorkerCount` responsible for controlling no of concurrent tasks
- added signposts for better tracing
- removed `Transcriber` protocol, made `WhipserKit` an open class
- added `TranscribeTask` class responsible for transcribing audio chunk to text and moved all the logic there
- added tests

This is how signposts look like in Instruments (for processing 5 audio files):

Screenshot 2024-04-04 at 11 30 49

Some benchmarks on my MacBook Air M1 (running in the release mode using tiny model `time swift run -c release whisperkit-cli transcribe [...]`):
- running on 1 file, 40:26 length -- 99.30s user, 11.07s system, 1:24.12 total
- running on 5 files, 40:26 length each -- 744.40s user, 111.61s system, 2:58.57 total
- running on 1 file, 0:11 length -- 0.73s user, 0.11s system, 1.535 total
- running on 5 files, 0:11 length each -- 2.55s user, 0.40s system, 1.780 total

using `Alice.mp3` file provided by @ZachNagengast:
- running on 1 file, 12:16 length -- 32.69s user, 3.71s system, 27.981 total
- running on 5 files, 12:16 length each -- 263.17s user, 39.25s system, 1:00.95 total

#### `WhisperKit`

1.

Deprecated

```swift
public func transcribe(
    audioPath: String,
    decodeOptions: DecodingOptions? = nil,
    callback: TranscriptionCallback = nil
) async throws -> TranscriptionResult?
```

use instead

```swift
public func transcribe(
    audioPath: String,
    decodeOptions: DecodingOptions? = nil,
    callback: TranscriptionCallback = nil
) async throws -> [TranscriptionResult]
```

2.

Deprecated

```swift
public func transcribe(
    audioArray: [Float],
    decodeOptions: DecodingOptions? = nil,
    callback: TranscriptionCallback = nil
) async throws -> TranscriptionResult?
```

use instead

```swift
public func transcribe(
    audioArray: [Float],
    decodeOptions: DecodingOptions? = nil,
    callback: TranscriptionCallback = nil
) async throws -> [TranscriptionResult]
```

#### `TextDecoding`

1.

Deprecated

```swift
func decodeText(
    from encoderOutput: MLMultiArray,
    using decoderInputs: DecodingInputs,
    sampler tokenSampler: TokenSampling,
    options decoderOptions: DecodingOptions,
    callback: ((TranscriptionProgress) -> Bool?)?
) async throws -> [DecodingResult]
```

use instead

```swift
func decodeText(
    from encoderOutput: MLMultiArray,
    using decoderInputs: DecodingInputs,
    sampler tokenSampler: TokenSampling,
    options decoderOptions: DecodingOptions,
    callback: ((TranscriptionProgress) -> Bool?)?
) async throws -> DecodingResult
```

2.

Deprecated

```swift
func detectLanguage(
    from encoderOutput: MLMultiArray,
    using decoderInputs: DecodingInputs,
    sampler tokenSampler: TokenSampling,
    options: DecodingOptions,
    temperature: FloatType
) async throws -> [DecodingResult]
```

use instead

```swift
func detectLanguage(
    from encoderOutput: MLMultiArray,
    using decoderInputs: DecodingInputs,
    sampler tokenSampler: TokenSampling,
    options: DecodingOptions,
    temperature: FloatType
) async throws -> DecodingResult
```

### Breaking changes

- removed `Transcriber` protocol

#### `AudioProcessing`

```swift
static func loadAudio(fromPath audioFilePath: String) -> AVAudioPCMBuffer?
```

becomes

```swift
static func loadAudio(fromPath audioFilePath: String) throws -> AVAudioPCMBuffer
```

#### `AudioStreamTranscriber`

```swift
public init(
    audioProcessor: any AudioProcessing, 
    transcriber: any Transcriber, 
    decodingOptions: DecodingOptions, 
    requiredSegmentsForConfirmation: Int = 2, 
    silenceThreshold: Float = 0.3, 
    compressionCheckWindow: Int = 20, 
    useVAD: Bool = true, 
    stateChangeCallback: AudioStreamTranscriberCallback?
)
```

becomes

```swift
public init(
    audioEncoder: any AudioEncoding,
    featureExtractor: any FeatureExtracting,
    segmentSeeker: any SegmentSeeking,
    textDecoder: any TextDecoding,
    tokenizer: any WhisperTokenizer,
    audioProcessor: any AudioProcessing,
    decodingOptions: DecodingOptions,
    requiredSegmentsForConfirmation: Int = 2,
    silenceThreshold: Float = 0.3,
    compressionCheckWindow: Int = 20,
    useVAD: Bool = true,
    stateChangeCallback: AudioStreamTranscriberCallback?
)
```

#### `TextDecoding`

```swift
func prepareDecoderInputs(withPrompt initialPrompt: [Int]) -> DecodingInputs?
```

becomes

```swift
func prepareDecoderInputs(withPrompt initialPrompt: [Int]) throws -> DecodingInputs
```

### CODE REVIEW

File: Sources/WhisperKit/Core/WhisperKit.swift [hunk 1]
```diff
@@ -8,15 +8,24 @@ import Foundation
...
 public protocol Transcriber {
+    func transcribe(audioPaths: [String], decodeOptions: DecodingOptions?, callback: TranscriptionCallback) async -> [String: Result<TranscriptionResult, Swift.Error>]
+    func transcribe(audioArrays: [[Float]], decodeOptions: DecodingOptions?, callback: TranscriptionCallback) async -> [Result<TranscriptionResult, Swift.Error>]
     func transcribe(audioPath: String, decodeOptions: DecodingOptions?, callback: TranscriptionCallback) async throws -> TranscriptionResult?
     func transcribe(audioArray: [Float], decodeOptions: DecodingOptions?, callback: TranscriptionCallback) async throws -> TranscriptionResult?
 }
 
 @available(macOS 13, iOS 16, watchOS 10, visionOS 1, *)
-public class WhisperKit: Transcriber {
+public actor WhisperKit: Transcriber {
```
Comment (ZachNagengast) [inline Sources/WhisperKit/Core/WhisperKit.swift:28]: Curious, why this change?

File: Sources/WhisperKit/Core/Utils.swift [hunk 2]
```diff
@@ -558,3 +566,10 @@ public class Logging {
         }
     }
 }
+
+
+enum Constants {
+    enum Logging {
+        static let subsystem = "com.argmax.whisperkit.Whisper"
+    }
+}
```
Comment (ZachNagengast) [inline Sources/WhisperKit/Core/Utils.swift:575]: Id put this in `Models.swift`

File: Sources/WhisperKit/Core/Utils.swift [hunk 3]
```diff
@@ -122,6 +122,14 @@ extension String {
...
 }
 
+extension Array {
+    func chunked(into size: Int) -> [[Element]] {
+        return stride(from: 0, to: count, by: size).map {
+            Array(self[$0 ..< Swift.min($0 + size, count)])
+        }
+    }
+}
+
```
Comment (ZachNagengast) [inline Sources/WhisperKit/Core/Utils.swift:132]: Id put this in `Models.swift` in the extensions section

**Review by ZachNagengast — COMMENTED** [review]
Prediction side of this looks great, nicely done. I left a few comments on how it could be cleaned up a bit. 

There is one missing part that is up to you if you want to include in this PR or add later, which is the actual chunking strategy for a continuous audio array. We will want to use the VAD and select breakpoints as close to 30s that does not cut off any speech in progress, with perhaps a threshold for how much silence to aim for to make our cuts (lots of experimentation to do here). Note that if you choose to do that in a later PR, then some of the comments can be skipped below, and this can simply be multiple file batching. Without a chunking strategy that uses VAD then we wouldn't want to chunk single files at all because it would degrade accuracy.

File: Sources/WhisperKitCLI/CLIArguments.swift [hunk 4]
```diff
@@ -4,8 +4,8 @@
 import ArgumentParser
 
 struct CLIArguments: ParsableArguments {
-    @Option(help: "Path to audio file")
-    var audioPath: String = "Tests/WhisperKitTests/Resources/jfk.wav"
+    @Option(help: "Paths to audio files")
+    var audioPath = [String]()
```
[the remark this answers is not in the document]
Comment (jkrukowski) [reply Sources/WhisperKitCLI/CLIArguments.swift:8]: yes, you can point to zero or more audio paths, e.g.

```
swift run whisperkit-cli transcribe --model-path openai_whisper-tiny --audio-path input1.mp3 --audio-path input2.mp3 --audio-path input3.mp3
```

File: Sources/WhisperKit/Core/WhisperKit.swift (hunk 1 above)
Comment (jkrukowski) [reply Sources/WhisperKit/Core/WhisperKit.swift:28]: `WhisperKit` contains shared mutable state (e.g. `currentTimings` property), since this PR introduces possibility to run inference on multiple files in parallel, mutating this state from different threads might lead to crashes. Wdyt?

Comment (jkrukowski) [reply Sources/WhisperKit/Core/WhisperKit.swift:28]: we could possibly make `TranscriptionTimings` an actor so it can be updated from the multiple threads and leave `WhisperKit` as a class

**Comment by jkrukowski** [comment]
> Prediction side of this looks great, nicely done. I left a few comments on how it could be cleaned up a bit.
> 
> There is one missing part that is up to you if you want to include in this PR or add later, which is the actual chunking strategy for a continuous audio array. We will want to use the VAD and select breakpoints as close to 30s that does not cut off any speech in progress, with perhaps a threshold for how much silence to aim for to make our cuts (lots of experimentation to do here). Note that if you choose to do that in a later PR, then some of the comments can be skipped below, and this can simply be multiple file batching. Without a chunking strategy that uses VAD then we wouldn't want to chunk single files at all because it would degrade accuracy.

I'd rather leave VAD for a separate PR if that's ok

File: Sources/WhisperKit/Core/TranscribeTask.swift [hunk 5]
```diff
@@ -0,0 +1,349 @@
...
+    private let audioEncoder: any AudioEncoding
+    private let featureExtractor: any FeatureExtracting
+    private let segmentSeeker: any SegmentSeeking
+    private let textDecoder: any TextDecoding
+    private let tokenizer: any WhisperTokenizer
+
+    init(
+        maxTokenContext: Int,
+        currentTimings: TranscriptionTimings,
+        progress: Progress?,
```
Comment (jkrukowski) [inline Sources/WhisperKit/Core/TranscribeTask.swift:29]: `Progress` [should be thread safe](https://oleb.net/blog/2014/03/nsprogress/) so I think we can pass it from the upstream

File: Sources/WhisperKit/Core/TranscribeTask.swift [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+                    throw WhisperError.transcriptionFailed("Encoder output is nil")
+                }
+                let encoderTime = Date().timeIntervalSince(encoderStart)
+                timings.encoding += encoderTime
+                timings.totalEncodingRuns += 1
+
+                // All features are computed, now we can decode
+                Logging.info("Decoding \(timeOffset)s - \(timeOffsetEnd)s")
+
+                try Task.checkCancellation()
```
Comment (jkrukowski) [inline Sources/WhisperKit/Core/TranscribeTask.swift:168]: added task cancellation checks

File: Sources/WhisperKit/Core/WhisperKit.swift [hunk 7]
```diff
@@ -8,15 +8,24 @@ import Foundation
...
+)
+private let signposter = OSSignposter(logger: logger)
 
 @available(macOS 13, iOS 16, watchOS 10, visionOS 1, *)
 public protocol Transcriber {
+    func transcribe(audioPaths: [String], decodeOptions: DecodingOptions?, callback: TranscriptionCallback) async -> [String: Result<TranscriptionResult, Swift.Error>]
+    func transcribe(audioArrays: [[Float]], decodeOptions: DecodingOptions?, callback: TranscriptionCallback) async -> [Result<TranscriptionResult, Swift.Error>]
     func transcribe(audioPath: String, decodeOptions: DecodingOptions?, callback: TranscriptionCallback) async throws -> TranscriptionResult?
     func transcribe(audioArray: [Float], decodeOptions: DecodingOptions?, callback: TranscriptionCallback) async throws -> TranscriptionResult?
 }
```
[the remark this answers is not in the document]
Comment (jkrukowski) [reply Sources/WhisperKit/Core/WhisperKit.swift]: removed Transcriber protocol, made WhipserKit an open class, everything goes through `func transcribe(audioArray:)` now

File: Sources/WhisperKit/Core/WhisperKit.swift [hunk 8]
```diff
@@ -385,26 +394,92 @@ public class WhisperKit: Transcriber {
...
+                        }
+                    }
+                }
+                var batchResult = [String: Result<TranscriptionResult, Swift.Error>]()
+                for await result in taskGroup {
+                    batchResult.merge(result) { $1 }
+                }
+                return batchResult
+            }
+            result.merge(partialResult) { $1 }
```
[the remark this answers is not in the document]
Comment (jkrukowski) [reply Sources/WhisperKit/Core/WhisperKit.swift]: everything goes through `func transcribe(audioArray:)` now

File: Sources/WhisperKit/Core/WhisperKit.swift (hunk 1 above)
Comment (jkrukowski) [reply Sources/WhisperKit/Core/WhisperKit.swift]: solved by introducing `TranscribeTask` class which is responsible for transcribing single chunk of audio

File: Sources/WhisperKit/Core/WhisperKit.swift [hunk 9]
```diff
@@ -418,426 +465,36 @@ public class WhisperKit: Transcriber {
...
-        var allSegments: [TranscriptionSegment] = []
-        var allTokens: [Int] = []
-        var transcription = ""
-
-        guard let tokenizer = tokenizer else {
+    public func transcribe(
+        audioArray: [Float],
+        decodeOptions: DecodingOptions? = nil,
+        callback: TranscriptionCallback = nil
+    ) async throws -> [TranscriptionResult] {
```
Comment (ZachNagengast) [inline Sources/WhisperKit/Core/WhisperKit.swift:472]: Breaking change, however in this case lets use a convenience method just above that has the same interface as before, and simply adjust the inputs and outputs to fit. This should be marked with a deprecated flag "Subject to removal in a future version, use {method signature} instead."

File: Tests/WhisperKitTests/TestUtils.swift [hunk 10]
```diff
@@ -234,3 +234,41 @@ extension SpecialTokens {
...
+    var text: String {
+        map(\.text).joined(separator: " ")
+    }
+}
+
+extension Collection where Element == TranscriptionResult {
+    var segments: [TranscriptionSegment] {
+        flatMap(\.segments)
+    }
+}
```
Comment (ZachNagengast) [inline Tests/WhisperKitTests/TestUtils.swift:274]: Aren't some of these needed for the result handling in core library? If so, perhaps these should be in the extension section as well

File: Sources/WhisperKit/Core/WhisperKit.swift [hunk 11]
```diff
@@ -371,42 +359,101 @@ public class WhisperKit: Transcriber {
...
     }
 
-    // MARK: - Transcribe audio file
+    // MARK: - Transcribe multiple audio files
+
+    public func transcribe(
+        audioPaths: [String],
+        decodeOptions: DecodingOptions? = nil,
+        callback: TranscriptionCallback = nil
+    ) async -> [Result<[TranscriptionResult], Swift.Error>] {
```
Comment (ZachNagengast) [inline Sources/WhisperKit/Core/WhisperKit.swift:380]: I support moving to `Result` types, but I think it should be gradual so as not to be jarring compared to the other single audio interfaces. Can you add support for both optional throwing and non-optional result formats of this interface, similar to the other convenience methods?

**Review by ZachNagengast — CHANGES_REQUESTED** [review]
Really exciting progress on this 🎉  the speed increase is _very_ impressive. In its current state it will obviously break a few interfaces when folks update to this version, so I wanted to make sure that was addressed by:
1. Convenience methods for core interfaces with breaking changes, which will allow developers to upgrade and still use the old methods, but still be able to upgrade to the new methods at their own pace. Marking them with a deprecated flag will show up in their warnings without causing errors.
Examples: 
`@available(*, deprecated, message: "Will be removed in future versions. Use transcribe(audioArrays:decodeOptions:callback:) instead.")`
or 
`@available(*, deprecated, renamed: "transcribe(audioArrays:decodeOptions:callback:)")`
2. An "Upgrade guide" for interfaces that have changes but are not getting backward compatible convenience methods. This would entail a short sentence about how the old interface compares to the new interface, and what to do when they get the error that would show up upon upgrading (if any).


Left a few sporadic comments as well (open to feedback on) and that should be the last remaining pieces to tidy up and then its good to go 👍

File: Sources/WhisperKit/Core/AudioStreamTranscriber.swift [hunk 12]
```diff
@@ -36,22 +36,36 @@ public actor AudioStreamTranscriber {
...
-    private let transcriber: any Transcriber
     private let decodingOptions: DecodingOptions
 
     public init(
+        maxTokenContext: Int,
+        audioEncoder: any AudioEncoding,
+        featureExtractor: any FeatureExtracting,
+        segmentSeeker: any SegmentSeeking,
+        textDecoder: any TextDecoding,
+        tokenizer: any WhisperTokenizer,
```
[the remark this answers is not in the document]
Comment (jkrukowski) [reply Sources/WhisperKit/Core/AudioStreamTranscriber.swift:49]: since the `Transcriber` is removed this will be a breaking change anyway, will prepare a doc with deprecation and breaking changes

File: Sources/WhisperKit/Core/AudioStreamTranscriber.swift [hunk 13]
```diff
@@ -194,11 +206,11 @@ public actor AudioStreamTranscriber {
         }
     }
 
-    private func transcribeAudioSamples(_ samples: [Float]) async throws -> TranscriptionResult? {
+    private func transcribeAudioSamples(_ samples: [Float]) async throws -> TranscriptionResult {
```
[the remark this answers is not in the document]
Comment (jkrukowski) [reply Sources/WhisperKit/Core/AudioStreamTranscriber.swift:209]: this is a private function so no risk of a breaking change here. Since it's a throwing function which returns the result of `TranscribeTask.run` which is non optional I'd leave as non optional

File: Tests/WhisperKitTests/TestUtils.swift (hunk 10 above)
Comment (jkrukowski) [reply Tests/WhisperKitTests/TestUtils.swift:274]: it's needed just for tests so I'd keep it here for now

File: Sources/WhisperKit/Core/AudioEncoder.swift [hunk 14]
```diff
@@ -2,6 +2,13 @@
 //  Copyright © 2024 Argmax, Inc. All rights reserved.
 
 import CoreML
+import OSLog
+
+private let logger = Logger(
+    subsystem: Constants.Logging.subsystem,
+    category: "AudioEncoding"
+)
+private let signposter = OSSignposter(logger: logger)
```
[the remark this answers is not in the document]
Comment (jkrukowski) [reply Sources/WhisperKit/Core/AudioEncoder.swift]: moved to `Logging`

File: Sources/WhisperKit/Core/AudioProcessor.swift [hunk 15]
```diff
@@ -21,8 +21,13 @@ public struct AudioDevice: Identifiable, Hashable {
 public protocol AudioProcessing {
     /// Loads audio data from a specified file path.
     /// - Parameter audioFilePath: The file path of the audio file.
-    /// - Returns: An optional `AVAudioPCMBuffer` containing the audio data.
-    static func loadAudio(fromPath audioFilePath: String) -> AVAudioPCMBuffer?
+    /// - Returns: `AVAudioPCMBuffer` containing the audio data.
+    static func loadAudio(fromPath audioFilePath: String) throws -> AVAudioPCMBuffer
```
[the remark this answers is not in the document]
Comment (jkrukowski) [reply Sources/WhisperKit/Core/AudioProcessor.swift:25]: added API changes in the PR description

File: Sources/WhisperKit/Core/AudioStreamTranscriber.swift (hunk 12 above)
[the remark this answers is not in the document]
Comment (jkrukowski) [reply Sources/WhisperKit/Core/AudioStreamTranscriber.swift]: + added API changes in the PR description

File: Sources/WhisperKit/Core/Models.swift [hunk 16]
```diff
@@ -1241,3 +1349,21 @@ extension WhisperTokenizerWrapper {
...
+
+// MARK: - Extensions
+
+extension Array {
+    func chunked(into size: Int) -> [[Element]] {
+        return stride(from: 0, to: count, by: size).map {
+            Array(self[$0 ..< Swift.min($0 + size, count)])
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (jkrukowski) [reply Sources/WhisperKit/Core/Models.swift]: moved

File: Sources/WhisperKit/Core/TextDecoder.swift [hunk 17]
```diff
@@ -34,15 +34,15 @@ public protocol TextDecoding {
...
+    ) async throws -> DecodingResult
     
     func detectLanguage(
         from encoderOutput: MLMultiArray,
         using decoderInputs: DecodingInputs,
         sampler tokenSampler: TokenSampling,
         options: DecodingOptions,
         temperature: FloatType
-    ) async throws -> [DecodingResult]
+    ) async throws -> DecodingResult
```
[the remark this answers is not in the document]
Comment (jkrukowski) [reply Sources/WhisperKit/Core/TextDecoder.swift]: added API changes in the PR description, added conviniece method

File: Sources/WhisperKit/Core/TextDecoder.swift [hunk 18]
```diff
@@ -720,7 +711,7 @@ open class TextDecoder: TextDecoding, WhisperMLModel {
             avgLogProb: avgLogProbs
         )
 
-        let decodingResult = DecodingResult(
+        return DecodingResult(
```
[the remark this answers is not in the document]
Comment (jkrukowski) [reply Sources/WhisperKit/Core/TextDecoder.swift]: changed

File: Sources/WhisperKit/Core/Models.swift [hunk 19]
```diff
@@ -414,9 +455,76 @@ public enum WhisperError: Error, LocalizedError {
 
 public struct TranscriptionResult: Codable {
     public var text: String
+    public var totalTokens: Int
```
[the remark this answers is not in the document]
Comment (jkrukowski) [reply Sources/WhisperKit/Core/Models.swift]: removed

File: Sources/WhisperKit/Core/WhisperKit.swift [hunk 20]
```diff
@@ -418,426 +465,36 @@ public class WhisperKit: Transcriber {
...
-        var allTokens: [Int] = []
-        var transcription = ""
-
-        guard let tokenizer = tokenizer else {
+    public func transcribe(
+        audioArray: [Float],
+        decodeOptions: DecodingOptions? = nil,
+        callback: TranscriptionCallback = nil
+    ) async throws -> [TranscriptionResult] {
+        precondition(modelState == .loaded, "Model has to be loaded at this point")
```
[the remark this answers is not in the document]
Comment (jkrukowski) [reply Sources/WhisperKit/Core/WhisperKit.swift]: changed to throwing an error

File: Sources/WhisperKit/Core/WhisperKit.swift (hunk 11 above)
Comment (jkrukowski) [reply Sources/WhisperKit/Core/WhisperKit.swift:396]: added additional method without `Result`

File: Examples/WhisperAX/WhisperAX/Views/ContentView.swift [hunk 21]
```diff
@@ -1186,12 +1183,14 @@ struct ContentView: View {
...
             return nil
         }
 
-        let transcription = try await whisperKit.transcribe(audioArray: samples, decodeOptions: options, callback: decodingCallback)
-        return transcription
+        return try await whisperKit.transcribe(
+            audioArray: samples,
+            decodeOptions: options,
+            callback: decodingCallback
+        ).first
```
[the remark this answers is not in the document]
Comment (ZachNagengast) [reply Examples/WhisperAX/WhisperAX/Views/ContentView.swift:1193]: This now has a deprecation warning 👍

**Review by ZachNagengast — APPROVED** [review]
Great work here, looks good to merge 👍  There is still some work to do for this bulk use case but this will be a great foundation to build on.
